zstd.cc

```diff
@@ -132,6 +136,82 @@ size_t zstd_processor::compress(const char* input, size_t input_len, char* outpu
     return ret;
 }
 
+template<typename Func, typename... Args>
+static size_t z_checked(std::string_view what, Func&& f, Args ...args) {
+    auto ret = f(args...);
+    if (ZSTD_isError(ret)) {
+        throw std::runtime_error(fmt::format("ZSTD {} failure: {}", what, ZSTD_getErrorName(ret)));
+    }
+    return ret;
+}
+
+size_t zstd_processor::uncompress(bytes_source& in, bytes_ostream& os) const {
+    z_checked("decompression", ZSTD_initDStream, _dctx.get());
+
+    size_t off = os.size_bytes();
+    ZSTD_outBuffer output = { nullptr, 0, 0 };
+
+    while (!in.empty()) {
+        auto frag = in.next();
+        auto size = frag.size();
+
+        ZSTD_inBuffer input = { frag.data(), size, 0 };
+
+        while (input.pos < input.size) {
```

Comment:
I think there's a bug here. `in.empty() && input.pos < input.size` is not a sufficient finish condition. Even after all input is ingested, the compressor might still have some output buffered. To drain all output correctly, you must keep calling `ZSTD_decompressStream` until you observe `output.pos < output.size`.

Reply:
Oh, okay. AFAICS this behavior is undocumented outside of that comment, that's why I didn't know about it. The manual (http://facebook.github.io/zstd/zstd_manual.html#Chapter9) says to check for `output.pos == output.size`. Maybe because the latter also works for infinite streams/frames.

Reply:
To be fair, the example could be wrong as well, but I hope not, since this way the code has a nicer control flow. :-/

Reply:
No, the example is right. The library has special code for making this property true, here: https://github.com/facebook/zstd/blob/86ead9f4a444a1017c951dd8681df7a284e0b54c/lib/decompress/zstd_decompress.c#L2348-L2365. The last input byte is "held hostage" until the output is fully flushed.

I'd guess they added this because it was a common user mistake not to check that the output is flushed.

Reply:
Woho.